R/generate_dust.R

```diff
@@ -675,6 +687,17 @@ generate_dust_assignment <- function(eq, name_state, dat, options = list()) {
                            idx$name, from, idx$name, to, idx$name),
                    res,
                    "}")
+        } else {          # idx$type == "single"
+          
+          # We need to replace [i,j,k....] on the right hand side with
+          # the matching var (or value) in the array[a,4,b] on the left.
+
+          lhs_index <- which(unlist(lapply(eq$lhs$array, function(x) {
+            x$name == idx$name})))
+          lhs_var <- eq$lhs$array[[lhs_index]]$at
+          eq$rhs$expr <- array_rewrite_index(eq$rhs$expr, idx$name, lhs_var)
+          rhs <- generate_dust_sexp(eq$rhs$expr, dat$sexp_data, options)
+          res <- sprintf("%s = %s;", lhs, rhs)
```

Comment:
This was the bit I was hoping you would look at the code generated by odin/odin.dust, please change to the implementation as used there, which is much simpler:

```suggestion
        } else {
          at <- generate_dust_sexp(idx$at, dat$sexp_data, options)
          res <- c("{",
                   sprintf("const size_t %s = %s;", idx$name, at),
                   res,
                   "}")
```

this also mirrors the range clause much more closely.

You will need to add a bit into the indentation code just below too:

```cc
      if (length(res) > 1) {
        i_start <- c(FALSE, grepl("^(for|\\{)", res[-length(res)]))
        i_end <- grepl("^}", res)
        indent <- strrep("  ", cumsum(i_start - i_end))
        res <- paste0(indent, res)
      }
```

replacing the block below